src/lib.rs

```diff
@@ -1,36 +1,74 @@
-use std::ptr::null_mut;
-use std::os::raw::c_int;
+mod libxenevtchn;
+extern crate xenevtchn_sys;
+use libxenevtchn::LibXenEvtchn;
+use log::debug;
+use std::convert::TryInto;
 use std::io::Error;
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+use xenevtchn_sys::{evtchn_port_t, xenevtchn_handle, xenevtchn_port_or_error_t};
 
 #[derive(Debug)]
 pub struct XenEventChannel {
-    handle: *mut xenevtchn_sys::xenevtchn_handle,
-    fd: c_int,
+    handle: *mut xenevtchn_handle,
     bind_port: c_int,
+    libxenevtchn: LibXenEvtchn,
 }
 
 impl XenEventChannel {
+    pub fn new(domid: u32, evtchn_port: u32) -> Result<Self, Error> {
+        let handle = unsafe { xenevtchn_sys::xenevtchn_open(null_mut(), 0) };
+        if handle == null_mut() {
+            return Err(Error::last_os_error());
+        }
+        let bind_port =
+            unsafe { xenevtchn_sys::xenevtchn_bind_interdomain(handle, domid, evtchn_port) };
+        debug!("bind_port = {x}", x = bind_port);
+        if bind_port < 0 {
+            return Err(Error::last_os_error());
+        }
 
-    pub fn new(domid: u32, evtchn_port: u32) -> Result<Self,Error> {
-        let handle = unsafe {
-            xenevtchn_sys::xenevtchn_open(null_mut(), 0)
-        };
-        if handle == null_mut() { return Err(Error::last_os_error()); }
-
-        let fd = unsafe {
-            xenevtchn_sys::xenevtchn_fd(handle)
-        };
-        let bind_port = unsafe {
-            xenevtchn_sys::xenevtchn_bind_interdomain(handle, domid, evtchn_port)
-        };
-        if bind_port < 0 { return Err(Error::last_os_error()); }
+        let libxenevtchn = unsafe { LibXenEvtchn::new() };
 
         Ok(XenEventChannel {
             handle,
-            fd,
             bind_port,
+            libxenevtchn,
         })
     }
+
+    pub fn xenevtchn_pending(&self) -> Result<xenevtchn_port_or_error_t, Error> {
+        let port = (self.libxenevtchn.xenevtchn_pending)(self.handle);
+        if port < 0 {
+            return Err(Error::last_os_error());
+        }
+        Ok(port)
+    }
+
+    pub fn get_bind_port(&self) -> Result<u32, Error> {
+        Ok(self.bind_port.try_into().unwrap())
```

Comment:
`c_int` is a signed integer, but you are converting it to an unsigned integer here. You should keep in mind that `c_int` is just a type alias for `i32` so you can simple write:
```suggestion
    pub fn get_bind_port(&self) -> Result<i32, Error> {
        Ok(self.bind_port)
```

Reply:
We are sure that bind_port will be positve. If it is negative it is an error and we have handled it in the new function